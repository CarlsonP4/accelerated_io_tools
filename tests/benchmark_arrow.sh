#!/bin/bash

## Run with: ./benchmark_arrow.sh <SIZE>
## SIZE is the sample data size in MB, default is 50MB

set -o errexit

DIR=`dirname $0`
IQ="iquery --afl --query"
F=/dev/shm/1

MB=$1
RE='^[0-9]+$'
if ! [[ $MB =~ $RE ]] ; then
   MB=50
fi

CNT=$((MB * 32768))            # MB * 1024 * 1024 / 8 / 4


echo "I. int64, int64, int64, int64"
echo "1. store"
$IQ "set no fetch;
     store(
       apply(
         build(
           <z:int64 not null>[i=1:$CNT],
           random()),
         y, int64(random()),
         x, int64(random()),
         w, int64(random())),
       foo)"

echo "2. summary"
$IQ "show(foo)"
$IQ "summarize(foo)"

echo "3. save"
time $IQ "save(foo, '$F', 0, '(int64,int64,int64,int64)')"
sz=`stat --printf="%s" $F`
echo "$((sz / 1024 / 1024)) MB ($sz B)"

echo "4. aio_save(binary)"
time $IQ "aio_save(foo, '$F', 'format=(int64,int64,int64,int64)')"
sz=`stat --printf="%s" $F`
echo "$((sz / 1024 / 1024)) MB ($sz B)"

echo "5. aio_save(arrow)"
time $IQ "aio_save(foo, '$F', 'format=arrow')"
sz=`stat --printf="%s" $F`
echo "$((sz / 1024 / 1024)) MB ($sz B)"

echo "6. Arrow read"
time python -c "import pyarrow; pyarrow.open_stream('$F').read_pandas()"

echo "7. SciDB-Py fetch"
time python -c "import scidbpy; scidbpy.connect().arrays.foo.fetch(atts_only=True)"

echo "8. SciDB-Py fetch w/ Arrow"
time python -c "import scidbpy; scidbpy.connect().arrays.foo.fetch(atts_only=True, use_arrow=True)"

$IQ "remove(foo)"


echo "II. int64, int64, int64, int64, string('')"
echo "1. store"
$IQ "set no fetch;
     store(
       apply(
         build(
           <z:int64 not null>[i=1:$CNT],
           random()),
         y, int64(random()),
         x, int64(random()),
         w, int64(random()),
         v, ''),
       foo)"

echo "2. summary"
$IQ "show(foo)"
$IQ "summarize(foo)"

echo "3. save"
time $IQ "save(foo, '$F', 0, '(int64,int64,int64,int64,string)')"
sz=`stat --printf="%s" $F`
echo "$((sz / 1024 / 1024)) MB ($sz B)"

echo "4. aio_save(binary)"
time $IQ "aio_save(foo, '$F', 'format=(int64,int64,int64,int64,string)')"
sz=`stat --printf="%s" $F`
echo "$((sz / 1024 / 1024)) MB ($sz B)"

echo "5. aio_save(arrow)"
time $IQ "aio_save(foo, '$F', 'format=arrow')"
sz=`stat --printf="%s" $F`
echo "$((sz / 1024 / 1024)) MB ($sz B)"

echo "6. Arrow read"
time python -c "import pyarrow; pyarrow.open_stream('$F').read_pandas()"

echo "7. SciDB-Py fetch"
time python -c "import scidbpy; scidbpy.connect().arrays.foo.fetch(atts_only=True)"

echo "8. SciDB-Py fetch w/ Arrow"
time python -c "import scidbpy; scidbpy.connect().arrays.foo.fetch(atts_only=True, use_arrow=True)"

$IQ "remove(foo)"
