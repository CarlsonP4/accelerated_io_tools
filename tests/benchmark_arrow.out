16384000
I. int64, int64, int64, int64
1. store
Query was executed successfully
2. summary
{i} schema
{0} 'foo<z:int64 NOT NULL,y:int64 NOT NULL,x:int64 NOT NULL,w:int64 NOT NULL> [i=1:16384000:0:1000000]'
{inst,attid} att,count,bytes,chunks,min_count,avg_count,max_count,min_bytes,avg_bytes,max_bytes
{0,0} 'all',16384000,524293712,85,384000,963765,1000000,48,6.16816e+06,8000072
3. save
{i} z,y,x,w

real	0m6.111s
user	0m0.008s
sys	0m0.007s
500 MB (524288000 B)
4. aio_save(binary)
{chunk_no,dest_instance_id,source_instance_id} val

real	0m2.392s
user	0m0.008s
sys	0m0.006s
500 MB (524288000 B)
5. aio_save(arrow)
{chunk_no,dest_instance_id,source_instance_id} val

real	0m3.415s
user	0m0.008s
sys	0m0.004s
500 MB (524289428 B)
6. Arrow read

real	0m0.560s
user	0m0.498s
sys	0m0.488s
7. SciDB-Py fetch

real	0m7.386s
user	0m1.939s
sys	0m1.753s
8. SciDB-Py fetch w/ Arrow

real	0m7.115s
user	0m1.708s
sys	0m1.824s
Query was executed successfully
II. int64, int64, int64, int64, string('')
1. store
Query was executed successfully
2. summary
{i} schema
{0} 'foo<z:int64 NOT NULL,y:int64 NOT NULL,x:int64 NOT NULL,w:int64 NOT NULL,v:string NOT NULL> [i=1:16384000:0:1000000]'
{inst,attid} att,count,bytes,chunks,min_count,avg_count,max_count,min_bytes,avg_bytes,max_bytes
{0,0} 'all',16384000,524295038,102,384000,963765,1000000,48,5.14015e+06,8000072
3. save
{i} z,y,x,w,v

real	0m9.533s
user	0m0.009s
sys	0m0.002s
578 MB (606208000 B)
4. aio_save(binary)
{chunk_no,dest_instance_id,source_instance_id} val

real	0m3.269s
user	0m0.011s
sys	0m0.006s
578 MB (606208000 B)
5. aio_save(arrow)
{chunk_no,dest_instance_id,source_instance_id} val

real	0m4.726s
user	0m0.011s
sys	0m0.004s
562 MB (589825764 B)
6. Arrow read

real	0m0.763s
user	0m0.644s
sys	0m0.536s
7. SciDB-Py fetch

real	12m37.844s
user	11m36.864s
sys	0m12.024s
8. SciDB-Py fetch w/ Arrow

real	0m10.977s
user	0m2.053s
sys	0m2.062s
Query was executed successfully
